src/classes/ai/aiHelpers.js

```diff
@@ -0,0 +1,88 @@
+/**
+ * @file aiHelpers.js file
+ * @author Steven on 2020-06-11
+ *
+ * A collection of helper functions for comparisons and other general tasks
+ * related to AI decision making.
+ */
+
+
+/**
+ * Comparator for sorting stacks to play a variable card.
+ * Expects that all stacks will accept a variable card already.
+ */
+function varStackCompare (a, b) {
+  // sort by stacks with Rx and best score
+  if (a.getTop().type === "REPEAT" && b.getTop().type === "REPEAT") {
+    return b.getScore() - a.getScore()
+  } else if (a.getTop().type === "REPEAT") {
+    return -1
+  } else if (b.getTop().type === "REPEAT") {
+    return 1
+  }
+
+  // sort by stack with lowest variable card
+  return lowestVar(a) - lowestVar(b) 
+}
+
+/**
+ * Helper to find the lowest variable in a stack.
+ * Expects there to be at least one variable in the stack.
+ * @return the value of the lowest variable card.
+ */
+function lowestVar (stack) {
+  let cards = stack.cards.filter(c => c.type === "VARIABLE")
+  return cards.sort((a,b) => { return a.value - b.value }).shift().value
+}
+
+/**
+ * Attempts to find a subset of stacks with scores that sum to value.
+ * Uses a heuristic that picks random stacks and adding them to the subset
+ * until the goal is reached or exceeded.
+ * The number of picks attempts to be high enough to find a subset for
+ * most reasonable values in our case, but low enough to guarantee that
+ * it will still run in less that 0.1 sec.
+ * This does not guarantee that a group will be found if one exists. We can
+ * veiw this as the AI player not being perfect and sometimes missing the
+ * right choice. In short games where there are very few stacks it should
+ * almost always find a grouping.
+ * The stacks given should always have scores <= value to make it easier
+ * for the algorithm to find a good subset.
+ * @param value The size of the group we are making.
+ * @param stacks The stacks that are available to group.
+ * @return a Set of stacks that group to make the value, or undefined if
+ * no grouping is possible.
+ */
+function findGroup (value, stacks) {
+  let MAX = 1000  // may be able to go as high as 10,000
```

Comment:
I wonder if we can have this threshold as a parameter to have AIs with different experience levels, as the idea here is "how deep to search the tree". Something to think about.

Reply:
There are a number of "AI" search tree algorithms (e.g. A* - https://en.wikipedia.org/wiki/A*_search_algorithm) that may be useful here. Something that could be explored in the future.

Reply:
I fixed this in another branch. It now uses dynamic programming and will always find a subset with that uses the most cards if one exists. Ie) group 6 with stack scores [1, 1, 1, 3, 3, 4] will return [1,1,1,3] and not [1,1,4] or [3,3].

Though I believe overall the way I constructed the AI we could find ways to use different group strategies to be more or less successful at finding groups. It should be as simple as sub classing the PlayBestCardAction and overriding specific methods, or creating specific action objects to deal with certain cards in different ways.

I have removed the random algorithm, but it could be re-introduced later. It is pretty simple.

Reply:
Actually looks like I did that work on this branch before creating the new branch, so fixing reviews added the new algorithm. Since I have unit tested it and confirmed it works I will just merge it here and we can make additions with other algorithms in the future if desired.